### ISSUE TITLE
Autofill/custom field property

### ISSUE THREAD

**Issue by trevorr** [erikras/redux-form#848]
Consider a postal address form where typing in a postal code can (in some cases) auto-fill the city and state/province fields. If the postal code is changed, it seems the auto-fill should only occur (again) if the city/state fields were not modified by the user (in other words, if they're blank or the previously auto-filled value). However, I'm not sure how to determine this.

- `dirty` and `pristine` seem to be relative to the initial value, so an auto-filled field would be considered `dirty`.
- `touched` and `visited` don't seem relevant, as tabbing through the fields shouldn't make them permanently un-auto-fillable.
- Component state can't be reliably used since fields can migrate across component instances (e.g. when an element of an array in a deep form is removed).

Therefore, it seems like a new field property is necessary. Is it possible/supported to create custom field properties? It seems like a reducer plugin could (and would be necessary to) accomplish this, but I'd want to make sure that redux-form wouldn't inadvertently drop it in some cases. It might also be nice to have explicit support for this, similar to how change actions can currently update the `touched` property.

**Comment by erikras** [comment]
So you're pining for a boolean flag like, `wasEverChangedByUser`?

Or, to be more succinct, let's call it `modified`. And then, you want, presumably via an async call, to be able to fire off an ajax request when `postalCode` changes `&& !city.modified && !state.modified`, to look up `city` and `state` values and set them?

That's both a totally reasonable use case and totally beyond the scope of most form libraries.

**Comment by trevorr** [comment]
`modified` is totally beyond the scope of most form libraries? That statement seems somewhat hyperbolic at least. :-)

What I'm really asking is whether the library would/could permit and propagate a custom field property in its state. I actually just finished trying it out, and the only problem was that the custom property was not propagated from the form state to the component props. Here's the reducer plugin:

```
const autofillReducer = (next: ?Reducer) =>
  (state: Object = {}, action: { type: string }): Object => {
    if (action.type === FormAction.AUTOFILL) {
      const { field, value } = action
      state = write(field + '.autofill', value, state)
      state = write(field + '.value', value, state)
    }
    return next ? next(state, action) : state
  }
```

The `autofill` property shows up in the state correctly, but doesn't make it to the form props.

**Comment by davidkpiano** [comment]
> That's both a totally reasonable use case and totally beyond the scope of most form libraries.

There actually might be a straightforward way to accomplish this. In RRF, the [`actions.load()`](https://davidkpiano.gitbooks.io/react-redux-form/content/model_actions.html#actionsloadmodel-value) action does the same thing as `actions.change()`, except that it is _silent_ - it completely bypasses the form reducer (and thus the form state never gets changed).

What this means is an async autofill function could `dispatch(actions.load('city', 'Miami'))`, which would update the city in the model state and in the view, but the form would still think that it was `.pristine`.

That way, all you have to do is check if it's `.pristine` before autofilling again :)

**Comment by erikras** [comment]
> `modified` is totally beyond the scope of most form libraries? That statement seems somewhat hyperbolic at least. :-)

😄  `most !== redux-form`

> The `autofill` property shows up in the state correctly, but doesn't make it to the form props.

Perhaps if there were some special key in the form state, like `meta` or something, where anything put there would be copied into the field object in the form props...

In `v6`, the form state is separate, so a special key wouldn't even be needed.

@trevorr Let me try to restate your feature request to make sure I'm understanding.

* A new action type: `autoFill(field, value)` that
  * Updates the value of a field if and only if the current value `===` the initial value (i.e. that field is `pristine`) OR the current value `===` the previously autofilled value
  * The field object in the props gets an `autofilled` boolean that is `true` when the current value `===` the autofilled value.
  * `pristine` and `dirty` remain relative to `initialValue`

**Comment by trevorr** [comment]
@davidkpiano Yes, if `dirty`/`pristine` are only determined by the change event (e.g. as opposed to comparison with `initialValue` at an unspecified time), that could be an elegant way to achieve this. How does that idea sound to you, @erikras?

@erikras Well, that would be the most "library does it all for me" approach. It implies that the autofilled value is tracked with the field as well, so in my attempted implementation, I just had the `autofill` property be the autofilled value rather than a boolean. That way, `change()` doesn't have to know to clear boolean. The other difference is that rather than putting the whether-to-autofill logic in the `autofill()` action, I made the action "dumb" and put the logic in the app. That has at least two benefits:

- If autofill shouldn't occur, any work necessary to obtain the autofilled value can be avoided. In my case, this would be the postal code lookup. Not a big deal, but slightly more efficient.
- When correlated fields are involved, like city and state, it probably only makes sense to autofill both or neither. (At least, that's what my app does currently.) Only the app would know about the correlation, so it seems like the logic would need to be in it.

As for `v6`, I'm excited about the performance improvement, but I haven't fully evaluated/internalized it yet, since I rely heavily on "deep" forms (which are still TBD there AFAIK). "Deep" forms are also the reason I couldn't cheat here and put autofill state in the component, since nested form state objects can migrate across component instances (due to `removeField`).

**Comment by erikras** [comment]
> if dirty/pristine are only determined by the change event (e.g. as opposed to comparison with initialValue at an unspecified time), that could be an elegant way to achieve this.

That seems like a misuse of that flag. I think `dirty` should mean "!== the initial value", which is the standard definition I've seen used in other libraries.

> The other difference is that rather than putting the whether-to-autofill logic in the autofill() action, I made the action "dumb" and put the logic in the app.

Your bullet points make great sense. So the library could just provide an `autofilledValue` from the last time it was autofilled, and that would be cleared when a `CHANGE` action modified the field.

> I rely heavily on "deep" forms (which are still TBD there AFAIK)

Deep fields are supported via the dot syntax, e.g. `<Field name="contact.billing.street"/>`, but _array fields_ are not yet supported. So if you are maintaining arrays of fields in `v5`, `v6` is not yet for you.

**Comment by davidkpiano** [comment]
> I think dirty should mean "!== the initial value", which is the standard definition I've seen used in other libraries.

Which other libraries? According to Angular (with by far the largest use of `dirty` in reference to forms):

> A model is considered to be dirty when the control has been changed from when first compiled.

This means that pristine/dirty should follow this state machine:

```
-> [pristine] ---(CHANGE or SET_DIRTY)---> [dirty]
              <---(SET_PRISTINE)----------
```

which it does in this visual example: https://scotch.io/tutorials/angularjs-form-validation

**Comment by trevorr** [comment]
> So the library could just provide an autofilledValue from the last time it was autofilled, and that would be cleared when a CHANGE action modified the field.

Well, if you did want to make `CHANGE` aware of the autofill property and clear it, a boolean could work just as well.  Then the autofill check becomes something like `pristine || autofilled`, which seems both clearer and more efficient than value comparisons. So to summarize the changes:

- Add an `autofilled` property to all fields that defaults to `false`.
- Add an `AUTOFILL` action that acts like `CHANGE`, except it sets `autofilled` to `true` (and does nothing with `touched`).
- Modify the `CHANGE` action to set `autofilled` to `false`.

If that sounds good, I can try it out and send you a PR. I'm still using 4.2.0, so I'd start with that branch.

BTW, do you have any thoughts on how apps should dispatch `AUTOFILL`? Starting with the action creator and plumbing it through via `mapDispatchToProps` is a bit of work, as it also requires binding the form and field names at some point. Would it be reasonable to add an `autofill()` method to the field properties, similar to `onChange()`?

> array fields are not yet supported. So if you are maintaining arrays of fields in v5, v6 is not yet for you.

Ah, yes, I'm using the hell out of _array_ fields. I've got arbitrary numbers of addresses, phone numbers, email addresses, payment methods, invoice line items, etc. I've definitely had to keep an eye  on performance/re-rendering, since the forms can be fairly large.

**Comment by erikras** [comment]
@davidkpiano, I must be remembering from ExtJS, where if you add a character to a value and then erase it again, the field is pristine again. [Example](http://jsfiddle.net/sran/SgkzZ/).

So in Angular, a `dirty` form control can only ever be made pristine again by specifically resetting it? Interesting. Can you provide any arguments as to why that is better than `=== initValue`? My way lets you know if you actually have to send the data back to the server or not.

---

@trevorr 

> * Add an autofilled property to all fields that defaults to false.

No need to default it to `false`. `undefined` is fine. `CHANGE` can `delete` it, and when it comes time to put the state into the props, you just `!!` it to make it a boolean. Seems cleaner than to muck up the state with a feature that most forms won't ever use.

> If that sounds good, I can try it out and send you a PR. I'm still using 4.2.0, so I'd start with that branch.

If by "that branch", you mean `master`, so that the feature would appear in a `v5.x` release, then sure.

> Would it be reasonable to add an `autofill()` method to the field properties, similar to `onChange()`?

That's only extra credit if you want to. Leaving it just as a `dispatch(autofill(form, field, value))` API seems acceptable.

**Comment by erikras** [comment]
Published as [`v5.2.0`](https://github.com/erikras/redux-form/releases/tag/v5.2.0).
